include/range/v3/view/set_algorithm.hpp

```diff
@@ -0,0 +1,904 @@
+        
+        
+        
+        namespace detail
+        {
+            template<bool IsConst,
+                     typename Rng1, typename Rng2,
+                     typename C, typename P1, typename P2>
+            struct set_intersection_cursor
+                : detail::set_op_cursor_move<set_intersection_cursor<IsConst, Rng1, Rng2, C, P1, P2>>
+            {
+            private:
+                using pred_ref_ = semiregular_ref_or_val_t<function_type<C>, IsConst>;
+                using proj1_ref_ = semiregular_ref_or_val_t<function_type<P1>, IsConst>;
+                using proj2_ref_ = semiregular_ref_or_val_t<function_type<P2>, IsConst>;
+                pred_ref_ pred_;
+                proj1_ref_ proj1_;
+                proj2_ref_ proj2_;
+                
+                template<typename T>
+                using constify_if = meta::invoke<meta::add_const_if_c<IsConst>, T>;
+                
+                using R1 = constify_if<Rng1>;
+                using R2 = constify_if<Rng2>;
+
+                range_iterator_t<R1> it1_;
+                range_sentinel_t<R1> end1_;
+                
+                range_iterator_t<R2> it2_;
+                range_sentinel_t<R2> end2_;
+                
+                void satisfy()
+                {
+                    while(it1_ != end1_ && it2_ != end2_)
+                    {
+                        if(pred_(proj1_(*it1_), proj2_(*it2_)))
+                            ++it1_;
+                        else
+                        {
+                            if(!pred_(proj2_(*it2_), proj1_(*it1_)))
+                                return;
+
+                            ++it2_;
+                        }
+                    }
+                }
+                
+            public:
+                using value_type = range_value_t<R1>;
+
+                set_intersection_cursor() = default;
+                set_intersection_cursor(pred_ref_ pred, proj1_ref_ proj1, proj2_ref_ proj2,
+                                        range_iterator_t<R1> it1, range_sentinel_t<R1> end1,
+                                        range_iterator_t<R2> it2, range_sentinel_t<R2> end2)
+                  : pred_(std::move(pred)), proj1_(std::move(proj1)), proj2_(std::move(proj2)),
+                    it1_(std::move(it1)), end1_(std::move(end1)), it2_(std::move(it2)), end2_(std::move(end2))
+                {
+                    satisfy();
+                }
+                auto get() const
+                RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
+                (
+                    *it1_
+                )
+                void next()
+                {
+                    ++it1_;
+                    ++it2_;
+                    satisfy();        
+                }
+                bool equal(set_intersection_cursor const &that) const
+                {
+                    return (it1_ == that.it1_ && it2_ == that.it2_) ||
+                           (done() && that.done());
```

Comment:
How much handholding do we want here? Can we trust the user not to compare cursors pointing into different ranges?